cardano-api/src/Cardano/Api/IO.hs

```diff
@@ -0,0 +1,52 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE KindSignatures #-}
+
+module Cardano.Api.IO
+  ( File(..)
+  , Directory(..)
+  , fileMap
+  , In
+  , Out
+  , InOut
+  , HasFileMode(..)
+  ) where
+
+import           Data.Aeson (FromJSON, ToJSON)
+import           Data.Kind (Type)
+import           Data.String (IsString)
+
+-- | Phantom type to indicate a file for reading.
+data In
+
+-- | Phantom type to indicate a file for writing.
+data Out
+
+-- | Phantom type to indicate a file for reading or writing.
+data InOut
+
+-- | A file path with additional type information to indicate whether it is to
+-- be used for reading or writing.
+newtype File direction = File { unFile :: FilePath }
+  deriving (Eq, Ord)
+  deriving newtype (Show, IsString, FromJSON, ToJSON)
+
+-- | A directory path.
+newtype Directory = Directory { unDirectory :: FilePath }
+  deriving (Eq, Ord)
+  deriving newtype (Show, IsString, FromJSON, ToJSON)
+
+fileMap :: (FilePath -> FilePath) -> File direction -> File direction
+fileMap f = File . f . unFile
+
+class HasFileMode (f :: Type -> Type) where
+  usingIn :: f InOut -> f In
+  usingOut :: f InOut -> f Out
+
+instance HasFileMode File where
```

Comment:
I think we are abstracting prematurely here. We can define the functions:

```
 usingIn :: File InOut -> File In
  usingIn = File . unFile

  usingOut :: File InOut -> File Out
  usingOut = File . unFile
  
```
rather than introducing a typeclass